flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadePoolingImpl.java

```diff
@@ -0,0 +1,183 @@
+    private transient Deque<XaFacade> pooled;
+
+    XaFacadePoolingImpl(FacadeSupplier facadeSupplier) {
+        this.facadeSupplier = facadeSupplier;
+    }
+
+    @Override
+    public void open() throws Exception {
+        checkState(active == null);
+        active = facadeSupplier.get();
+        active.open();
+        pooled = new LinkedList<>();
+        prepared = new HashMap<>();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return active != null && active.isOpen();
+    }
+
+    @Override
+    public void start(Xid xid) throws TransientXaException {
+        active.start(xid);
+    }
+
+    @Override
+    public void endAndPrepare(Xid xid) throws Exception {
+        checkState(!prepared.containsKey(xid));
+        active.endAndPrepare(xid);
+        prepared.put(xid, active);
+        if (pooled.isEmpty()) {
+            active = facadeSupplier.get();
+            active.open();
+        } else {
+            active = pooled.poll();
+        }
+    }
+
+    @Override
+    public void commit(Xid xid, boolean ignoreUnknown) throws TransientXaException {
+        runForXid(xid, facade -> facade.commit(xid, ignoreUnknown));
+    }
+
+    @Override
+    public void rollback(Xid xid) throws TransientXaException {
+        runForXid(xid, facade -> facade.rollback(xid));
+    }
+
+    @Override
+    public void failOrRollback(Xid xid) throws TransientXaException {
+        runForXid(xid, facade -> facade.failOrRollback(xid));
+    }
+
+    @Override
+    public Collection<Xid> recover() throws TransientXaException {
+        return peekPooled().recover();
+    }
+
+    @Override
+    public void close() throws Exception {
+        active.close();
+        for (XaFacade facade : prepared.values()) {
+            facade.close();
+        }
+        for (XaFacade facade : pooled) {
+            facade.close();
+        }
+    }
+
+    @Nullable
+    @Override
+    public Connection getConnection() {
+        return active.getConnection();
+    }
+
+    @Override
+    public boolean isConnectionValid() throws SQLException {
+        return active.isConnectionValid();
+    }
+
+    @Override
+    public Connection getOrEstablishConnection() throws SQLException, ClassNotFoundException {
+        return active.getOrEstablishConnection();
+    }
+
+    @Override
+    public void closeConnection() {
+        active.closeConnection();
+    }
+
+    @Override
+    public Connection reestablishConnection() throws SQLException, ClassNotFoundException {
+        return active.reestablishConnection();
+    }
+
+    // WARN: action MUST leave the facade in IDLE state (i.e. not start/end/prepare any tx)
+    private void runForXid(Xid xid, ThrowingConsumer<XaFacade, TransientXaException> action) {
+        XaFacade mapped = prepared.remove(xid);
+        if (mapped == null) {
+            LOG.debug("No XA resource found associated with XID: {}", xid);
+            action.accept(peekPooled());
```

Comment:
when this would happen?

that says we have a prepared trans to commit/rollback/fail without a connection associated with it.

During the recovery I guess.

Document it a bit?

Reply:
Yes, this can happen on recovery, I'll add a comment.